com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs

```diff
@@ -17,37 +17,117 @@ public string name
+        {
+            m_ChildObjectList.Remove(itemToRemove);
+            if (m_ChildObjectIDSet.Contains(itemToRemove.objectId))
+                m_ChildObjectIDSet.Remove(itemToRemove.objectId);
+        }
+
+        public void MoveItemInCategory(ShaderInput itemToMove, int newIndex)
+        {
+            m_ChildObjectList.Remove(itemToMove);
+            m_ChildObjectList.Insert(newIndex,itemToMove);
+            // Recreate the hash-set as the data structure does not allow for moving/inserting elements
+            m_ChildObjectIDSet.Clear();
+            foreach (var childObject in m_ChildObjectList)
+            {
+                m_ChildObjectIDSet.Add(childObject.value.objectId);
+            }
+        }
+
+        public bool IsItemInCategory(ShaderInput itemToCheck)
+        {
+            return m_ChildObjectIDSet.Contains(itemToCheck.objectId);
+        }
 
-        HashSet<Guid> m_ChildItemIDSet;
-        // We expose Guids as a HashSet for faster existence checks
-        public HashSet<Guid> childItemIDSet
+        public bool IsNamedCategory()
         {
-            get => m_ChildItemIDSet;
-            set => m_ChildItemIDSet = value;
+            return name != String.Empty;
         }
 
-        public CategoryData(string inName,  List<Guid> inChildItemIDList = null, Guid inCategoryGuid = new Guid())
+        public override void OnAfterDeserialize()
+        {
+            if (m_ChildObjectList != null)
+            {
+                foreach (var childObject in m_ChildObjectList.ToList())
+                {
+                    if (childObject.value != null)
+                        m_ChildObjectIDSet.Add(childObject.value.objectId);
+                    else
+                        m_ChildObjectList.Remove(childObject);
+                }
+            }
+
+            base.OnAfterDeserialize();
+        }
+
+        public CategoryData()
+        {
+            foreach (var childObject in m_ChildObjectList)
+            {
+                m_ChildObjectIDSet.Add(childObject.value.objectId);
+            }
+        }
+
+        public CategoryData(string inName, List<ShaderInput> categoryChildren = null)
         {
             name = inName;
-            m_ChildItemIDList = inChildItemIDList;
-            if (m_ChildItemIDList != null)
-                m_ChildItemIDSet = new HashSet<Guid>(m_ChildItemIDList);
-            else
-                AssertHelpers.Fail("Category data provided invalid data for construction.");
-            categoryGuid = inCategoryGuid;
+            if (categoryChildren != null)
+            {
+                foreach (var childObject in categoryChildren)
+                {
+                    m_ChildObjectList.Add(childObject);
+                    m_ChildObjectIDSet.Add(childObject.objectId);
+                }
+            }
+        }
+
+        public CategoryData(CategoryData categoryToCopy)
+        {
+            this.name = categoryToCopy.name;
```

Comment:
Shouldn't category copying also copy the objects?

Reply:
```
        public CategoryData CopyCategory(CategoryData categoryToCopy)
        {
            var copiedCategory = new CategoryData(categoryToCopy);
            AddCategory(copiedCategory);
            // Whenever a category is copied, also copy over all the inputs within that category
            foreach (var childInputToCopy in categoryToCopy.Children)
            {
                var newShaderInput = AddCopyOfShaderInput(childInputToCopy);
                copiedCategory.InsertItemIntoCategory(newShaderInput);
            }

            return copiedCategory;
        }
```

But looking at that function again, seeing why I chose to implement it like that, which is that we need to create a copy of the input first, and `AddCopyOfShaderInput` handles a decent amount of stuff that I didn't want to mess with or move from `GraphData`.